DigitalLearningSolutions.Web/Attributes/SetApplicationTypeAndSelectedTab.cs

```diff
@@ -0,0 +1,46 @@
+﻿namespace DigitalLearningSolutions.Web.Attributes
+{
+    using System;
+    using DigitalLearningSolutions.Data.Enums;
+    using DigitalLearningSolutions.Web.Models.Enums;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.AspNetCore.Mvc.Filters;
+
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method)]
+    public class SetApplicationTypeAndSelectedTab : Attribute, IActionFilter
+    {
+        private readonly string? applicationTypeName;
+
+        private readonly string? tabName;
+
+        public SetApplicationTypeAndSelectedTab(string? applicationTypeName, string? tabName)
+        {
+            this.applicationTypeName = applicationTypeName;
+            this.tabName = tabName;
+        }
+
+        public void OnActionExecuted(ActionExecutedContext context) { }
+
+        public void OnActionExecuting(ActionExecutingContext context)
+        {
+            if (!(context.Controller is Controller controller))
+            {
+                return;
+            }
+
+            if (tabName != null)
+            {
+                controller.ViewData["SelectedTab"] = (Tab)tabName;
+            }
+
+            if (applicationTypeName == null)
+            {
+                controller.ViewData["ApplicationType"] = context.ActionArguments["application"];

```

Comment:
Is this to pick up the application for pages like the My account page? i.e. does it rely on the route parameter being called "application"?

Reply:
Should probably have some kind of documentation to explain this, I'm not sure whether there will be a better way to resolve this, but we should have something somewhere to explain why it might be failing with an incorrect name.

Reply:
https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/19822084230/Selected+tab+functionality

Reply:
Documentation is good, but I think in this case needing that documentation is a sign that the code can be improved.
I think we should actually have two attributes, one for setting the selected tab, and the other for setting the application.

Here's why:
* Configuring the application and configuring the selected tab do two quite different things - application sets the header text and nav bar, while selected tab just sets the selected-tab marker.
* Application can depend the url params, and selected tab can't
* Endpoints can have an application without having a selected tab, and vice versa.

The application endpoint can then have an additional `ApplicationRouteParamName` parameter, which you could set to get the attribute to run this line. And you can throw an exception in the constructor if both this param and a concrete application are given, because that conflicts.